streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/TaskManager.scala

```diff
@@ -18,141 +18,159 @@
+import akka.actor._
 import akka.pattern.ask
+import org.apache.gearpump.TimeStamp
 import org.apache.gearpump.cluster.AppMasterContext
 import org.apache.gearpump.cluster.MasterToAppMaster.ReplayFromTimestampWindowTrailingEdge
 import org.apache.gearpump.cluster.scheduler.{Resource, ResourceAllocation}
 import org.apache.gearpump.streaming.AppMasterToExecutor.{LaunchTask, StartClock}
 import org.apache.gearpump.streaming.Executor.RestartExecutor
 import org.apache.gearpump.streaming.ExecutorToAppMaster.RegisterTask
-import org.apache.gearpump.streaming.TaskSet.TaskLaunchData
+import org.apache.gearpump.streaming.TaskSchedulerImpl.TaskLaunchData
 import org.apache.gearpump.streaming.appmaster.AppMaster.AllocateResourceTimeOut
-import org.apache.gearpump.streaming.appmaster.ExecutorManager.{SetTaskManager, StartExecutors}
+import org.apache.gearpump.streaming.appmaster.ExecutorManager._
+import org.apache.gearpump.streaming.storage.InMemoryAppStoreOnMaster
 import org.apache.gearpump.streaming.task._
-import org.apache.gearpump.streaming.{DAG, TaskSet}
-import org.apache.gearpump.transport.HostPort
+import org.apache.gearpump.streaming.{TaskScheduler, DAG, TaskSchedulerImpl}
 import org.apache.gearpump.util.{ActorUtil, Constants, LogUtil}
 import org.slf4j.Logger
 
 import scala.concurrent.Future
+import TaskManager._
 
 private[appmaster] class TaskManager(
-    appContext: AppMasterContext,
+    appId: Int,
     dag: DAG,
-    clockService: ActorRef,
+    taskScheduler: TaskScheduler,
     executorManager: ActorRef,
+    clockService: ActorRef,
     appMaster: ActorRef)
-  extends Actor {
+  extends Actor with FSM[State, StateData] {
 
-  import appContext.appId
-private val LOG: Logger = LogUtil.getLogger(getClass, app = appId)
+  private val LOG: Logger = LogUtil.getLogger(getClass, app = appId)
+  val systemConfig = context.system.settings.config
 
   implicit val timeout = Constants.FUTURE_TIMEOUT
   implicit val actorSystem = context.system
-
   import context.dispatcher
-  private var taskLocations = Map.empty[HostPort, Set[TaskId]]
-  private var executorToTaskIds = Map.empty[Int, Set[TaskId]]
-  private var startedTasks = Set.empty[TaskId]
 
-  private val taskSet = new TaskSet(appId, dag)
-  //When the AppMaster trying to replay, the replay command should not be handled again.
-  private var restarting = true
+  startWith(Uninitialized, null)
+  self ! DagInit(dag)
 
-  executorManager ! SetTaskManager(self)
-  LOG.info("Sending request resource to master...")
-  val resourceRequests = taskSet.fetchResourceRequests()
-  executorManager ! StartExecutors(resourceRequests)
+  when (Uninitialized) {
+    case Event(DagInit(dag), _) =>
+      executorManager ! SetTaskManager(self)
+      LOG.info("Sending request resource to master...")
 
-  def receive = executorMessageHandler orElse taskMessageHandler orElse recoveryHandler
+      taskScheduler.setTaskDAG(dag)
+      val resourceRequests = taskScheduler.getResourceRequests()
```

Comment:
Because we have a side effect here, by 
```
//This has a side effect
setTaskDAG(dag)

//this has no side effect
getResourceRequests()
```
This declaration doesnot indicate it has a side effect: getResourceRequests(dag)
That is why in the beginning, I prefered name "dagUpdated" over getResourceRequests. But you made a good point that dagUpdated doesn't show what it is doing in this function.